tools/kata-webhook/main.go

```diff
@@ -0,0 +1,177 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/kata-containers/kata-containers/src/runtime/pkg/katautils"
+	"github.com/sirupsen/logrus"
+	kwhhttp "github.com/slok/kubewebhook/v2/pkg/http"
+	kwhlogrus "github.com/slok/kubewebhook/v2/pkg/log/logrus"
+	kwhmodel "github.com/slok/kubewebhook/v2/pkg/model"
+	kwhmutating "github.com/slok/kubewebhook/v2/pkg/webhook/mutating"
+)
+
+func getEnvValue(envKey, defaultEnvValue string) string {
+	if envValue, ok := os.LookupEnv(envKey); ok {
+		return envValue
+	}
+	return defaultEnvValue
+}
+
+func annotatePodMutator(_ context.Context, ar *kwhmodel.AdmissionReview, obj metav1.Object) (*kwhmutating.MutatorResult, error) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		// If not a pod just continue the mutation chain (if there is one) and don't do anything
+		return &kwhmutating.MutatorResult{}, nil
+	}
+
+	// The Namespace is not always available in the pod Spec
+	// specially when operators create the pods. Hence access
+	// the Namespace in the actual request (vs the object)
+	// https://godoc.org/k8s.io/api/admission/v1beta1#AdmissionRequest
+	if whPolicy.nsBlacklist[ar.Namespace] {
+		fmt.Println("blacklisted namespace: ", ar.Namespace)
+		return &kwhmutating.MutatorResult{}, nil
+	}
+
+	// We cannot support --net=host in Kata
+	// https://github.com/kata-containers/documentation/blob/master/Limitations.md#docker---nethost
+	if pod.Spec.HostNetwork {
+		fmt.Println("host network: ", pod.GetNamespace(), pod.GetName())
+		return &kwhmutating.MutatorResult{}, nil
+	}
+
+	if pod.GetNamespace() == "sonobuoy" {
+		fmt.Println("sonobuoy pods will not be changed to kata", pod.GetNamespace(), pod.GetName())
+		return &kwhmutating.MutatorResult{}, nil
+	}
+
+	for i := range pod.Spec.Containers {
+		if pod.Spec.Containers[i].SecurityContext != nil && pod.Spec.Containers[i].SecurityContext.Privileged != nil {
+			if *pod.Spec.Containers[i].SecurityContext.Privileged {
+				fmt.Println("privileged container: ", pod.GetNamespace(), pod.GetName())
+				return &kwhmutating.MutatorResult{}, nil
+			}
+		}
+	}
+
+	if pod.Spec.RuntimeClassName != nil {
+		fmt.Println("explicit runtime: ", pod.GetNamespace(), pod.GetName(), pod.Spec.RuntimeClassName)
+		return &kwhmutating.MutatorResult{}, nil
+	}
+
+	// Mutate the pod
+	fmt.Println("setting runtime to kata: ", pod.GetNamespace(), pod.GetName())
+
+	runtimeClassEnvKey := "RUNTIME_CLASS"
+	kataRuntimeClassName := getEnvValue(runtimeClassEnvKey, "kata")
+	pod.Spec.RuntimeClassName = &kataRuntimeClassName
+
+	if getEnvValue("MANAGE_RESOURCES", "true") == "true" {
+		fmt.Println("setting default limits and request values: ", pod.GetNamespace(), pod.GetName())
+
+		hypConfig := katautils.GetDefaultHypervisorConfig()
```

Comment:
Where to get cpu and memory limit from? Trying GetDefaultHypervisorConfig()

Reply:
After talking to @fidencio, we opted to take the default values from toml configuration and let the user define where is the configuration. For our current setup, is `/opt/confidential-containers/share/defaults/kata-containers`. We should take `default_maxvcpus` and `default_maxmemory` and use these for cpu and memory limit in case the user does not specifies in the config map

Reply:
I think the webhook should either define its own defaults/values or we should introduce new values in the config that designate default resources available for/to the containers (potentially similar to `default_memory` and `default_vcpus` which define the resources for the "UVM OS" part of the sandbox). `default_maxvcpus` and `default_maxmemory` does not sound appropriate here.

For now, I would suggest to define those values in the webhook. If the webhook gets upstreamed, one can attempt to change the main code base. The main code base could either take the resource limits e.g. set by the webhook, or defined by the user. If the webhook is not running and the user does not specify limits, default values could be used and added to the VM resources (e.g. UVM OS memory + container memory = memory size of UVM)

Reply:
Changing to https://github.com/kata-containers/kata-containers/pull/7304/commits/c2ef95baf2fb5fb771724a91223675cc128548ee for now and open to explore better solution